src/preprocessing/passes/bv_to_int.cpp

```diff
@@ -440,16 +440,38 @@ Node BVToInt::translateWithChildren(Node original,
         // translate the result to integers
         returnNode = d_nm->mkNode(kind::BITVECTOR_TO_NAT, bvand);
       }
-      else
+      else if (options::solveBVAsInt() == options::SolveBVAsIntMode::SUM)
       {
-        Assert(options::solveBVAsInt() == options::SolveBVAsIntMode::SUM);
         // Construct a sum of ites, based on granularity.
         Assert(translated_children.size() == 2);
         returnNode =
-            d_iandTable.createBitwiseNode(translated_children[0],
-                                          translated_children[1],
-                                          bvsize,
-                                          options::BVAndIntegerGranularity());
+            d_iandUtils.createSumNode(translated_children[0],
+                                      translated_children[1],
+                                      bvsize,
+                                      options::BVAndIntegerGranularity());
+      }
+      else
+      {
+        Assert(options::solveBVAsInt() == options::SolveBVAsIntMode::BITWISE);
+        uint64_t granularity = options::BVAndIntegerGranularity();
+
+        Node x = translated_children[0];
+        Node y = translated_children[1];
+        returnNode = d_nm->mkNode(kind::IAND, x, y);
+
+        // eagerly add bitwise lemmas according to the provided granularity
+        uint64_t high_bit;
+        for (uint64_t j = 0; j < bvsize; j += granularity)
+        {
+          high_bit = j + granularity - 1;
+          // don't let high_bit pass bvsize
+          if (high_bit >= bvsize)
+          {
+            high_bit = bvsize - 1;
+          }
+          d_bitwiseAssertions.insert(returnNode.eqNode(
+              d_iandUtils.createBitwiseIAndNode(x, y, high_bit, j)));
```

Comment:
It seems to me like maybe the entire loop can be implemented inside `createBitwiseIAndNode`, as u have a similar loop in the `iand` solver. Is it possible to push more into that function?

Reply:
Yeah that's a good point. It's very similar, the only difference is that it's lazy in the `iand` solver so it would need to be able to check for each range of bits whether or not the abstraction needs to be refined. I'm not sure what the right interface would be for that, so I just left it outside.

Reply:
ok that makes sense. maybe then document it and explicitly mention that it is lazy and in what sense?

Reply:
Does this make sense? 33dca66629606eb6ecbf69077d5aedd2ed184080